x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionStringFieldMapper.java

```diff
@@ -0,0 +1,432 @@
+                            boolean accepted = runAutomaton.run(decoded, 0, decoded.length);
+                            if (accepted) {
+                                return AcceptStatus.YES;
+                            }
+                            return AcceptStatus.NO;
+                        }
+                    };
+                }
+            };
+        }
+
+        @Override
+        public Query wildcardQuery(String value, MultiTermQuery.RewriteMethod method, QueryShardContext context) {
+            if (context.allowExpensiveQueries() == false) {
+                throw new ElasticsearchException(
+                    "[wildcard] queries cannot be executed when '" + ALLOW_EXPENSIVE_QUERIES.getKey() + "' is set to false."
+                );
+            }
+            failIfNotIndexed();
+
+            VersionFieldWildcardQuery query = new VersionFieldWildcardQuery(new Term(name(), value));
+            QueryParsers.setRewriteMethod(query, method);
+            return query;
+        }
+
+        @Override
+        protected BytesRef indexedValueForSearch(Object value) {
+            String valueAsString;
+            if (value instanceof String) {
+                valueAsString = (String) value;
+            } else if (value instanceof BytesRef) {
+                // encoded string, need to re-encode
+                valueAsString = ((BytesRef) value).utf8ToString();
+            } else {
+                throw new IllegalArgumentException("Illegal value type: " + value.getClass() + ", value: " + value);
+            }
+            return encodeVersion(valueAsString).bytesRef;
+        }
+
+        @Override
+        public IndexFieldData.Builder fielddataBuilder(String fullyQualifiedIndexName, Supplier<SearchLookup> searchLookup) {
+            failIfNoDocValues();
+            return new SortedSetOrdinalsIndexFieldData.Builder(name(), VersionScriptDocValues::new, CoreValuesSourceType.BYTES);
+        }
+
+        @Override
+        public Object valueForDisplay(Object value) {
+            if (value == null) {
+                return null;
+            }
+            return VERSION_DOCVALUE.format((BytesRef) value);
+        }
+
+        @Override
+        public DocValueFormat docValueFormat(@Nullable String format, ZoneId timeZone) {
+            if (format != null) {
+                throw new IllegalArgumentException("Field [" + name() + "] of type [" + typeName() + "] does not support custom formats");
+            }
+            if (timeZone != null) {
+                throw new IllegalArgumentException(
+                    "Field [" + name() + "] of type [" + typeName() + "] does not support custom time zones"
+                );
+            }
+            return VERSION_DOCVALUE;
+        }
+
+        @Override
+        public Query rangeQuery(Object lowerTerm, Object upperTerm, boolean includeLower, boolean includeUpper, QueryShardContext context) {
+            if (context.allowExpensiveQueries() == false) {
```

Comment:
from your last benchmarking tests, `rangeQuery` doesn't seem to be expensive, so this check to allow expensive queries looks unnecessary to me. WDYT?

Reply:
While I'm happy removing this check, the same check exists for keyword fields via StringFieldType#rangeQuery() which is also using TermRangeQuery under the hood. While we expect the dictionary size for version fields to be lower, we cannot be sure users are not (by accident) misusing this field to store hundreds of thousands of version strings, in which case performance should be similar to keyword field performance which we currently guard for by checking "allowExpensiveQueries()". Wdyt?

Reply:
Thanks, I am also ok to have this check for consistency with `StringFieldType`.

Reply:
Unlike keyword, range queries are the most important part of this field so I'd prefer that we ignore the `allowExpensiveQueries` in this field.